ethereum/sidecar/server.go

```diff
@@ -428,6 +537,342 @@ func (s *Server) startGRPCServer(
+// to the highest).
+func (s *Server) fetchRecentAssetsUnlockedEvents(ctx context.Context) (
+	[]bridgetypes.AssetsUnlockedEvent,
+	error,
+) {
+	// Fetching events starts from the current sequence tip.
+	sequenceTip, err := s.bridgeOutClient.GetAssetsUnlockedSequenceTip(ctx)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"failed to fetch assets unlocked sequence tip: [%w]",
+			err,
+		)
+	}
+
+	// The sequence tip is zero, meaning no events have been made.
+	if sequenceTip.IsZero() {
+		return []bridgetypes.AssetsUnlockedEvent{}, nil
+	}
+
+	// Cut-off block time in UNIX seconds.
+	cutOffBlockTime := uint32(s.timeFunc().Add(-s.bridgeOutLookBackPeriod).Unix()) //nolint:gosec
+	recentEvents := []bridgetypes.AssetsUnlockedEvent{}
+
+	// Walk backwards from the current sequence tip in windows of at most
+	// `assetsUnlockedBatchSize`.
+outer:
+	for sequenceTip.IsPositive() {
+		// Sequence end is exclusive. We must add `1`.
+		seqEnd := sequenceTip.AddRaw(1)
+
+		// Sequence start is inclusive. It cannot be lower than `1`.
+		seqStart := seqEnd.SubRaw(int64(s.assetsUnlockedBatchSize))
+		if seqStart.LT(sdkmath.OneInt()) {
+			seqStart = sdkmath.OneInt()
+		}
+
+		events, err := s.bridgeOutClient.GetAssetsUnlockedEvents(
+			ctx,
+			seqStart,
+			seqEnd,
+		)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed to fetch AssetsUnlocked events [%s, %s): %w",
+				seqStart.String(),
+				seqEnd.String(),
+				err,
+			)
+		}
+
+		// We know how many events we should receive. Any difference suggests
+		// an error on `mezod` server side (e.g. gaps in stored event database).
+		expectedLength := seqEnd.Sub(seqStart).Int64()
+		if int64(len(events)) != expectedLength {
+			return nil, fmt.Errorf(
+				"fetched unexpected number of AssetsUnlocked events for "+
+					"range [%s, %s); expected %d, got %d",
+				seqStart.String(),
+				seqEnd.String(),
+				expectedLength,
+				len(events),
+			)
+		}
+
+		for i := len(events) - 1; i >= 0; i-- {
+			event := events[i]
+			if event.BlockTime < cutOffBlockTime {
+				// We found an event older than cut-off time.
+				break outer
+			}
+			recentEvents = append(recentEvents, event)
+		}
+
+		// If the length of fetched events was smaller than the batch size,
+		// there are no more events to fetch.
+		if len(events) < s.assetsUnlockedBatchSize {
+			break
+		}
+
+		sequenceTip = seqStart.SubRaw(1)
+	}
+
+	// Ensure events are sorted in ascending order of their unlock sequences.
+	sort.Slice(recentEvents, func(i, j int) bool {
+		return recentEvents[i].UnlockSequence.LT(recentEvents[j].UnlockSequence)
+	})
+
+	return recentEvents, nil
+}
+
+// findUnattestedAssetsUnlockedEvents finds unattested `AssetsUnlocked` events
+// from the passed input events. An event is considered attested if there
+// was an `AssetsUnlockConfirmed` event emitted for it on Ethereum and its
+// unlock sequence is among the confirmed unlocks.
+func (s *Server) findUnattestedAssetsUnlockedEvents(
```

Comment:
See previous comments:
```suggestion
func (s *Server) findUnconfirmedAssetsUnlockedEvents(
```
Let's update the docstring accordingly.

Reply:
Done in 0d2144781d4d683df2c171302b0c640049659133.